Bottom: 56639d9ce0870c67e476caa5dcc4d7a114ebd47f
Top:    61b795bc48b6ae582b8f6b40bff3c6dbd0000e04
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-13 00:43:51 +0000

common/hidgl.c: Support for drawing "fullpoly" polygons

This finally bit someone (and eneded up with them producing
a bad batch of boards), so it is about time I fixed this.


---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 73e85ce..4e167f6 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -613,7 +613,6 @@ static GLint stencil_bits;
 static int dirty_bits = 0;
 static int assigned_bits = 0;
 
-/* FIXME: JUST DRAWS THE FIRST PIECE.. TODO: SUPPORT FOR FULLPOLY POLYGONS */
 void
 hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box, double scale)
 {
@@ -664,6 +663,16 @@ hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box, double scale
   /* Drawing operations now set our reference bit in the stencil buffer */
 
   r_search (poly->Clipped->contour_tree, clip_box, NULL, do_hole, &info);
+  if (TEST_FLAG (FULLPOLYFLAG, poly))
+    {
+      PolygonType p = *poly;
+
+      for (p.Clipped = poly->Clipped->f;
+           p.Clipped != poly->Clipped;
+           p.Clipped = p.Clipped->f)
+        r_search (p.Clipped->contour_tree, clip_box, NULL, do_hole, &info);
+    }
+
   hidgl_flush_triangles (&buffer);
 
   glPopAttrib ();                               /* Restore the colour and stencil buffer write-mask etc.. */
@@ -680,6 +689,16 @@ hidgl_fill_pcb_polygon (PolygonType *poly, const BoxType *clip_box, double scale
 
   /* Draw the polygon outer */
   tesselate_contour (info.tobj, poly->Clipped->contours, info.vertices, scale);
+  if (TEST_FLAG (FULLPOLYFLAG, poly))
+    {
+      PolygonType p = *poly;
+
+      for (p.Clipped = poly->Clipped->f;
+           p.Clipped != poly->Clipped;
+           p.Clipped = p.Clipped->f)
+        tesselate_contour (info.tobj, p.Clipped->contours, info.vertices, scale);
+    }
+
   hidgl_flush_triangles (&buffer);
 
   /* Unassign our stencil buffer bit */
